File: Sujan/Microsoft/day2/que/largest_divisible_subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<int>v;
      
        int n=nums.size();
          vector<int>dp(n,1);
        int m=1,num=-1;
       for(int i=1;i<n;i++)
       {
           for(int j=0;j<i;j++)
           {
               if(!(nums[i]%nums[j]) && dp[i]<dp[j]+1)
               {
                   dp[i]=dp[j]+1;
                   if(m<dp[i])
                   {
                       m=dp[i];
                   }
               }
           }
       }
       for(int i=n-1;i>=0;i--)
       {
           if(m==dp[i] && (num==-1 || !(num%nums[i])))
           {
               v.push_back(nums[i]);
               m--;
               num=nums[i];
           }
       }
       sort(v.begin(),v.end());
  return v;
    }
};

File: Calto-GG/Microsoft/Day 2/Largest_Divisible_Subset.py
class Solution(object):
    def largestDivisibleSubset(self, nums):
        if not nums:
            return []
        nums.sort()
        # initialize the list dp with the same size as nums
        dp = [1] * len(nums)
        predecessor = [-1] * len(nums)
        max_size = 0
        max_index = -1
        for i, num in enumerate(nums):
            for j in range(i):
                if nums[i] % nums[j] == 0 and dp[i] < dp[j] + 1:
                    dp[i] = dp[j] + 1
                    predecessor[i] = j
            #update the maximum subset size and index if dp[i] is greater than the current maximum
            if dp[i] > max_size:
                max_size = dp[i]
                max_index = i
        # reconstruct the maximum divisible subset
        result = []
        while max_index >= 0:
            result.append(nums[max_index])
            max_index = predecessor[max_index]
        return result